forage-android/src/main/java/com/joinforage/forage/android/vault/BasisTheoryPinSubmitter.kt

```diff
@@ -66,51 +67,57 @@ internal class BasisTheoryPinSubmitter(
         return vaultToForageResponse(vaultResponse)
     }
 
-    override fun parseVaultError(vaultResponse: BasisTheoryResponse): String {
+    override fun parseVaultErrorMessage(vaultResponse: BasisTheoryResponse): String {
         return vaultResponse.exceptionOrNull().toString()
     }
 
     override fun toForageSuccessOrNull(vaultResponse: BasisTheoryResponse): ForageApiResponse.Success<String>? {
         // The caller should have already performed the error checks.
         // We add these error checks as a safeguard, just in case.
-        if (toVaultErrorOrNull(vaultResponse) != null ||
+        if (vaultResponse.isFailure ||
+            toVaultErrorOrNull(vaultResponse) != null ||
             toForageErrorOrNull(vaultResponse) != null
         ) {
             return null
         }
 
+        // note: Result.toString() wraps the actual response as
+        // "Success(<actual-value-here>)"
         return ForageApiResponse.Success(vaultResponse.getOrNull().toString())
     }
 
     override fun toForageErrorOrNull(vaultResponse: BasisTheoryResponse): ForageApiResponse.Failure? {
-        // is a Basis Theory error and doesn't have to do with Forage
-        if (vaultResponse.isFailure) return null
-
         return try {
-            // if the response is a ForageApiError, then this block
-            // should not throw
-            val forageResponse = vaultResponse.getOrNull().toString()
-            val forageApiError = ForageApiError.ForageApiErrorMapper.from(forageResponse)
-            val error = forageApiError.errors[0]
-            ForageApiResponse.Failure.fromError(
-                ForageError(400, error.code, error.message)
+            val matchedStatusCode = getBasisTheoryExceptionStatusCode(vaultResponse)!!
+            val basisTheoryExceptionBody = getBasisTheoryExceptionBody(vaultResponse)!!
+
+            val forageApiError = ForageApiError.ForageApiErrorMapper.from(basisTheoryExceptionBody)
+            val firstError = forageApiError.errors[0]
+
+            return ForageApiResponse.Failure.fromError(
+                ForageError(matchedStatusCode, firstError.code, firstError.message)
             )
-        } catch (_: JSONException) {
-            // if we throw when trying to extract the ForageApiError,
+        } catch (_: Exception) {
```

Comment:
The reason that we have always try-catched these functions is because the pattern you see in a lot of places where we try to parse to JSON (ex. `ForageApiError.ForageApiErrorMapper.from(basisTheoryExceptionBody)`) throws if the string doesn't conform.

Should we have two catch statements here? The first for `JSONException` and the second for all other exceptions so that we can handle them more carefully?

Reply:
The reason I use a regular `Exception` here is because the regex parsing may throw a `NullPointerException` (or maybe even an out-of-bounds exception) and the ForageApiErrorMapper may throw a `JSONException`

In any case, an exception in this block means that the vault response doesn't conform to `ForageApiError`

So instead of:

```kotlin
catch (_: JSONException) {
  null
} catch (_: NullPointerException) {
  null
} catch (_: IndexOutOfBoundsException) {
 null
}
```

I went with a more shorthand approach (`catch (_: Exception) {}`)

Reply:
You're right that the outcome would be your proposal above. Let's just go for it.